### DISCUSSION TITLE
What is the `Context` that is used by JoinSet::poll_join_next()?

### DISCUSSION THREAD

**Discussion by DanielT** [tokio-rs/tokio#6157] [Q&A] [answered]
I was just reading the documentation for tokio::JoinSet for the first time.
It contains the functions [poll_join_next()](https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html#method.poll_join_next) and [poll_join_next_with_id()](https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html#method.poll_join_next_with_id).

`poll_join_next()` looks like it would be useful to me, but the two functions are unlike all the others in JoinSet: they need a `Context`. The context is not created by any function I currently have in my code, neither its origin nor its use are explained in the documentation, and I have no idea how to get one.

Can someone please explain this to me here?

**Comment by Darksonn** [comment, marked as answer]
The `Future` trait is defined like this:
```rust
pub trait Future {
    type Output;
    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>;
}
```
[documentation](https://doc.rust-lang.org/stable/std/future/trait.Future.html)

So when you manually implement the `Future` trait, the context is available to you. It's not available when using the async/await syntax.

If all you want is a way try to get the next future if one is immediately available, then you shouldn't use the poll methods. Instead, use [`now_or_never`](https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.now_or_never) on the non-poll method. You may also want to file a feature request for a `try_join_next`. (Similar to how `mpsc::Receiver` has `recv`, `try_recv`, and `poll_recv`.)

**Reply by DanielT** [reply]
Thanks!
I hadn't looked "under the hood" of futures yet.
